Log Context: Add `cacheFilters` property

svennergr opened the pull request:
**What is this feature?**

This PR adds a `cacheFilters` property to the `getRowContextQuery` API. By default the property will always be `true` and thus, filters in e.g. Loki's LogContextProvider will always be cached.

However, when opening the context query in a new tab, these filters should not be cached, thus the API get's called with `cacheFilters` set to `false`.

svennergr committed:
add `forceApplyFilters` property

ivanahuckova commented on packages/grafana-data/src/types/logs.ts:
```diff
@@ -137,7 +137,12 @@ export interface DataSourceWithLogsContextSupport<TQuery extends DataQuery = Dat
+    row: LogRowModel,
+    options?: LogRowContextOptions,
+    query?: TQuery,
+    initialFilters?: boolean
```
```suggestion
    forceApplyFilters?: boolean
```

ivanahuckova replied:
I think this also need a comment explaining when to use `forceApplyFilters` and how is it currently used.

ivanahuckova commented on public/app/features/logs/components/LogRowMenuCell.tsx:
```diff
@@ -50,7 +54,7 @@ export const LogRowMenuCell = React.memo(
         ) {
           const win = window.open('about:blank');
-          const query = await getRowContextQuery(row);
+          const query = await getRowContextQuery(row, {}, true);
```
```suggestion
          const query = await getRowContextQuery(row, undefined, true);
```

As we are not adding any options, this could be also passed as undefined.

ivanahuckova commented on public/app/plugins/datasource/loki/LogContextProvider.ts:
```diff
@@ -53,12 +53,17 @@ export class LogContextProvider {
+  ) {
     const direction = (options && options.direction) || LogRowContextQueryDirection.Backward;
     const limit = (options && options.limit) || this.datasource.maxLines;
     // This happens only on initial load, when user haven't applied any filters yet
     // We need to get the initial filters from the row labels
```
We need to update a comment to include both scenarios:
- when `this.appliedContextFilters.length === 0` happens only on initial load, when user haven't applied any filters yet and in that case we need to get the initial filters from the row labels
- if `forceApplyFilters === true` we want to apply initial filters from the row labels (which is currently used for `cmd + click` on show context)

ivanahuckova commented on public/app/plugins/datasource/loki/LogContextProvider.ts:
```diff
@@ -75,15 +80,17 @@ export class LogContextProvider {
+    origQuery?: LokiQuery,
+    forceApplyFilters?: boolean
   ): Promise<LokiQuery> => {
     // FIXME: This is a hack to make sure that the context query is created with
     // the correct set of filters. The whole `appliedContextFilters` property
     // should be revisted.
```
```suggestion
```

ivanahuckova commented on public/app/plugins/datasource/loki/LogContextProvider.ts:
```diff
@@ -75,15 +80,17 @@ export class LogContextProvider {
+    const { query } = await this.getQueryAndRange(row, options, origQuery, forceApplyFilters);
+
+    if (forceApplyFilters) {
+      this.appliedContextFilters = [];
```
Also here, we need to add comment that explains that if `forceApplyFilters` is called, we need to reset filters after.

ivanahuckova reviewed (commented):
I think that this implementation make sense, but I am wondering if `forceApplyFilters` is a good name of prop and if it makes sense in context of other data sources.  The comment for `getLogRowContextQuery` is  `Retrieve the context query object for a given log row. This is currently used to open LogContext queries in a split view.` Maybe `forceApplyFilters` is a good name, but it needs a detailed comment. 

Plus as we are changing API + adding feature, we should also update (from https://docs.google.com/document/d/1knNsbtn0qnLtnp1gNazeZ6JV53Wh-BQeMf_oAF6_YYM/edit) :

### Documentation
- [x] Updated feature  documentation

### API change/addition
- [ ] Updated developers documentation
- [ ] Add new API to logs data source example

svennergr committed:
PR review

svennergr committed:
fix tests

svennergr committed:
remove import

svennergr committed:
comment flaky test

svennergr committed:
add context tab documentation

svennergr commented:
@ivanahuckova I think I addressed most of your feedback. I will do 

### API change/addition
- [ ] Updated developers documentation
- [ ] Add new API to logs data source example
 
in separate PRs in the corresponding repos.

ivanahuckova commented on docs/sources/explore/logs-integration.md:
```diff
@@ -130,6 +130,8 @@ You may encounter long lines of text that make it difficult to read and analyze
 
 The **Open in split view** button allows you to execute the context query for a log entry in a split screen in the Explore view. Clicking this button will open a new Explore pane with the context query displayed alongside the log entry, making it easier to analyze and understand the surrounding context.
 
+Log context can also be opened in a new browser tab by pressing the Cmd/Ctrl button while clicking on the button to open the context modal.
```
```suggestion
Log context query can also be opened in a new browser tab by pressing the Cmd/Ctrl button while clicking on the button to open the context modal.
```

ivanahuckova commented on packages/grafana-data/src/types/logs.ts:
```diff
@@ -136,8 +136,16 @@ export interface DataSourceWithLogsContextSupport<TQuery extends DataQuery = Dat
+   * @internal
    */
-  getLogRowContextQuery?: (row: LogRowModel, options?: LogRowContextOptions, query?: TQuery) => Promise<TQuery | null>;
+  getLogRowContextQuery?: (
```
I don't think this should be alpha/internal as we already have it in documentation https://grafana.com/developers/plugin-tools/tutorials/build-a-logs-data-source-plugin. So I would remove the `@alpha/@internal`. Or at least `@internal`. And also that documentation needs to be updated.

ivanahuckova commented on public/app/plugins/datasource/loki/LogContextProvider.ts:
```diff
@@ -363,8 +373,10 @@ export class LogContextProvider {
+        // Otherwise use new filters; also only show the notification if filters
+        // are supposed to be cached, which is currently used in the UI, not
+        // when tab-opened
+        if (arePreservedLabelsUsed && cacheFilters) {
```
I think that instead of this, we should not get to this logic and keep `preservedLabels` as false, as we are not using them. I don't think we even need to check local storage for preserved labels in that case. So we could do, even before checking for `preservedLabels`:
```
if (!cacheFilters) {
return contextFilters;
}
```

Also I noticed that we are not passing `cacheFilters` in LokiContextUi.

svennergr replied:
No, we are using them, it's just that we don't show the popup that filters are applied when the context is opened in a new tab.

svennergr replied:
> Also I noticed that we are not passing cacheFilters in LokiContextUi.

That's fine, as we always want to cache and the default is set to `true`.

svennergr committed:
review

svennergr committed:
doc
